tutorials/scripting/c_sharp/c_sharp_variant.rst

```diff
@@ -9,17 +9,67 @@ For a detailed explanation of Variant in general, see the :ref:`Variant <class_V
 We recommend avoiding ``Godot.Variant`` unless it is necessary to interact with untyped engine APIs.
 Take advantage of C#'s type safety when possible.
 
-Any of ``Variant.As{TYPE}`` methods or the generic ``Variant.As<T>`` method can be used to convert
-a ``Godot.Variant`` to a C# type. Since the ``Godot.Variant`` type contains implicit conversions
-defined for all the supported types, calling these methods directly is usually not necessary.
+Converting from a supported type to ``Godot.Variant`` can be done using implicit conversions. Also
+available are ``CreateFrom`` method overloads and the generic ``Variant.From<T>`` methods.
 
-Use ``CreateFrom`` method overloads or the generic ``Variant.From<T>`` method to convert a C# type
-to a ``Godot.Variant``.
+.. code-block:: csharp
+
+    int x = 42;
+    Godot.Variant number = x;
+    Godot.Variant hello = "Hello, World!";
+
+    Godot.Variant number2 = Godot.Variant.CreateFrom(x);
+    Godot.Variant number3 = Godot.Variant.From(x);
+
+Implicit conversions to ``Godot.Variant`` make passing variants as method arguments very convenient.
+For example, ``final_val`` of :ref:`tween_property<class_Tween_method_tween_property>`.
+
+.. code-block:: csharp
+
+    Tween tween = CreateTween();
+    tween.TweenProperty(GetNode("Sprite"), "modulate", Colors.Red, 1.0f);
+
+Converting from ``Godot.Variant`` to a supported type can be done using explicit conversions. Also
+available are ``Variant.As{TYPE}`` methods or the generic ``Variant.As<T>`` method.
+
+.. code-block:: csharp
+
+    int cSharpNumber = (int)number;
+    string cSharpHello = (string)hello;
+
+    int cSharpNumber2 = number.As<int>();
+    string cSharpHello2 = hello.AsString();
+
+.. note::
+
+    All methods of converting from ``Godot.Variant`` to a C# type behave the same way. If conversion
+    is not possible, the default value of the target type or an empty array is returned. An
+    exception is not thrown.
+
+A matching type is not necessary for a conversion to succeed. For example, all variants are
+convertible to ``string``.
+
+.. code-block:: csharp
+
+    string s = Variant.From(new Vector3(1, 3, 9)).As<string>();
+    GD.Print(s); // (1, 3, 9)
+
+Some C# types are not directly represented by ``Variant.Type``. The ``Variant.As{TYPE}`` method uses
+a more specific name to represent them.
+
+.. code-block:: csharp
+
+    int cSharpNumber3 = number.AsInt32();
+
+To convert ``Godot.Variant`` to a C# ``object`` without specifying a more specific target type, use
+the ``Variant.Obj`` property. This can be used to compare two variants that both have unknown types
+using the equality operator ``==``. When possible, prefer more specific conversions. ``Variant.Obj``
+may result in unnecessary boxing of value types.
```

Comment:
I don't know if it's a good idea to use `Variant.Obj` to compare Variant values, and in general I'd prefer to not encourage the usage of that property.

Using `==` with boxed values will fail:

```csharp
object a = 42;
object b = 42;
Console.WriteLine(a == b); // This prints False.
Console.WriteLine(a.Equals(b)); // This prints True.
```

It may be better to check the `VariantType` first, and then use the `As{TYPE}` or `As<T>` methods:

```csharp
public static bool IsEqualsTo(this Variant left, Variant right)
{
    if (left.VariantType != right.VariantType)
    {
        return false;
    }

    return switch left.VariantType
    {
        Variant.Type.Nil => true,
        Variant.Type.Bool => left.AsBool() == right.AsBool(),
        Variant.Type.Int => left.AsInt64() == right.AsInt64(),
        Variant.Type.Float => left.AsDouble() == right.AsDouble(),
        // ...
    }
}
```

Reply:
Hmm. I'll just remove this paragraph for now.

Adding it to the Variant API could raise questions like whether Int 1 == Float 1 (should it attempt to convert), or less likely Bool true, etc. Not sure what would be expected based on how GDScript/Godot behaves.

Reply:
Yeah I don't know either, I think it's a complicated topic where different users may have different needs (e.g.: JavaScript's `==` vs `===`). So I feel like it's better to leave it to a user-defined method.